giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+      double upperBound, double[][] xValues, double[] yValues)
+      throws Exception {
+
+      boolean result = false;
+      if (coefficient[coefIndex] < lowerBound ||
+        coefficient[coefIndex] > upperBound) {
+
+        double value;
+        if (coefficient[coefIndex] < lowerBound) {
+          value = lowerBound;
+        } else {
+          value = upperBound;
+        }
+        int ptr = -1;
+        if (validColumnIndices.size() >= 1 &&
+          validColumnIndices.get(validColumnIndices.size() - 1) == coefIndex) {
+          ptr = validColumnIndices.size() - 1;
+        } else if (validColumnIndices.size() >= 2 &&
+          validColumnIndices.get(validColumnIndices.size() - 2) == coefIndex) {
+          ptr = validColumnIndices.size() - 2;
+        }
+        if (ptr != -1) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("addRecord: coefficient at index " + coefIndex +
+              " is wrong in the regression, setting it to " + value);
+          }
+          // remove from valid column
+          validColumnIndices.remove(ptr);
+          // re-create the X matrix
+          fillXMatrix(dataSamples, validColumnIndices, xValues);
+          // adjust Y values
+          for (int i = 0; i < memorySamples.size(); ++i) {
+            yValues[i] -= value * dataSamples.get(i)[coefIndex];
+          }
+          // save new coefficient value in intermediate array
+          extreme[coefIndex] = value;
+          // re-run regression
+          mlr.newSampleData(yValues, xValues);
+          calculateRegression(coefficient, validColumnIndices, mlr);
+          result = true;
+        } else {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(
+              "addRecord: coefficient was not in the regression, " +
+                "setting it to the lower bound");
+          }
+          result = false;
+        }
+        coefficient[coefIndex] = value;
+      }
+      return result;
+    }
+
+    /**
+     * Calculates the regression.
+     * @param coefficient Array of coefficients
+     * @param validColumnIndices List of valid columns
+     * @param mlr {@link OLSMultipleLinearRegression} instance.
+     * @throws Exception
+     */
+    private static void calculateRegression(double[] coefficient,
+      Vector<Integer> validColumnIndices, OLSMultipleLinearRegression mlr)
+      throws Exception {
+
+      if (coefficient.length != validColumnIndices.size()) {
+        throw new Exception("There are " + coefficient.length +
+          " coefficients, but " + validColumnIndices.size() +
+          " valid columns in the regression");
+      }
+
+      double[] beta = mlr.estimateRegressionParameters();
+      for (int i = 0; i < coefficient.length; ++i) {
+        coefficient[i] = 0;
+      }
+      for (int i = 0; i < validColumnIndices.size(); ++i) {
+        coefficient[validColumnIndices.get(i)] = beta[i];
+      }
+      coefficient[5] = beta[validColumnIndices.size()];
```

Comment:
Are you sure that bias is the last parameter? I did this small test: 

`   
   OLSMultipleLinearRegression mlr = new OLSMultipleLinearRegression();
    double[] y = new double[] {1, 2, 3};
    double[][] x = new double[][] {{0}, {0.1}, {0.2}};
    mlr.newSampleData(y, x);
    System.out.println(Arrays.toString(mlr.estimateRegressionParameters()));
`
And it printed: [0.9999999999999997, 10.0] , so bias is actually on the position 0